File: 3743-reschedule-meetings-for-maximum-free-time-i/reschedule-meetings-for-maximum-free-time-i.cpp
class Solution {
public:
    int maxFreeTime(int eventTime, int k, vector<int>& startTime, vector<int>& endTime) {
        int n = startTime.size(),ans=0,t=0;


        vector<int> freeArray; 
        freeArray.push_back(startTime[0]);
        for(int i = 1; i < startTime.size(); i++)freeArray.push_back(startTime[i] - endTime[i-1]);
        freeArray.push_back(eventTime - endTime[endTime.size()-1]);

        int i=0,j=0,currSum=0,maxSum=0;
        // *max_element(freeArray.begin(),end(freeArray));
        
        if(k+1 >= freeArray.size()){
            currSum = accumulate(freeArray.begin(),end(freeArray),0);
            return currSum;
        }

        for(; j<=k; ++j)currSum += freeArray[j];
        maxSum=currSum;
        while(j<freeArray.size()){
            currSum += freeArray[j];
            currSum -= freeArray[i];
            i++;
            j++;
            maxSum = max(maxSum,currSum);
        }
        return maxSum;
    }
};

File: 3743-reschedule-meetings-for-maximum-free-time-i/reschedule-meetings-for-maximum-free-time-i.java
class Solution {
    public int maxFreeTime(int eventTime, int k, int[] startTime, int[] endTime) {
        int n = startTime.length;
        int[] freeTime = new int[n+1];
        freeTime[0]=startTime[0];
        for(int i=1; i<n; ++i)freeTime[i] = startTime[i]-endTime[i-1];
        freeTime[n] = eventTime-endTime[n-1];

        int l=0,currSum=0,maxSum=0;
        for(int r=0; r<freeTime.length; ++r){
            currSum += freeTime[r];
            if(r-l+1 > k+1){
                currSum -= freeTime[l];
                l++;
            }
            maxSum = Math.max(maxSum,currSum);
        }
        return maxSum;
    }
}